### PR TITLE
Pb 2279: Added fixes to take care new EncryptionV2Key variable in     backuplocation.

### PR DESCRIPTION
**Pull request by siva-portworx**
**What type of PR is this?**
>bug

**What this PR does / why we need it**:
pb-2279

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
yes
```release-note
Issue: Backups were not encrypted even though the encryption was pass in the backuplocation CR.
User Impact: Users were not able to take encrypted backup.
Resolution: Fixed to take the encryption key passed in the backuplocation CR and encrypt the backup.

```

**Does this change need to be cherry-picked to a release branch?**:
2.11 branch

### CODE REVIEW

**Review by diptiranjanpx — APPROVED** [review]

File: pkg/apis/stork/v1alpha1/backuplocation.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ type BackupLocationItem struct {
 	SecretConfig       string        `json:"secretConfig"`
 	Sync               bool          `json:"sync"`
 	RepositoryPassword string        `json:"repositoryPassword"`
+	EncryptionV2Key    string        `json:"encryptionV2Key"`
```
Comment (ram-infrac) [inline pkg/apis/stork/v1alpha1/backuplocation.go:46]: what's different between `EncryptionKey` and `EncryptionV2Key` ?

File: pkg/applicationmanager/controllers/applicationrestore.go [hunk 2]
```diff
@@ -721,10 +721,13 @@ func (a *ApplicationRestoreController) downloadObject(
 	if err != nil {
 		return nil, err
 	}
-	if restoreLocation.Location.EncryptionKey != "" {
-		if data, err = crypto.Decrypt(data, restoreLocation.Location.EncryptionKey); err != nil {
-			return nil, err
+	if restoreLocation.Location.EncryptionV2Key != "" {
+		var decryptData []byte
+		if decryptData, err = crypto.Decrypt(data, restoreLocation.Location.EncryptionV2Key); err != nil {
+			logrus.Errorf("ApplicationRestoreController/downloadObject: resource data decrypt failedi:%v, returing data direclty", err)
```
Comment (ram-infrac) [inline pkg/applicationmanager/controllers/applicationrestore.go:727]: nit: `decyrpt failed: %v, returning data directly`

File: pkg/apis/stork/v1alpha1/backuplocation.go (hunk 1 above)
Comment (siva-portworx) [reply pkg/apis/stork/v1alpha1/backuplocation.go:46]: EncryptionKey is deprecated now and we will use EncryptionV2Key for storing the encryption key. Added the comments in the definition.

File: pkg/applicationmanager/controllers/applicationrestore.go (hunk 2 above)
Comment (siva-portworx) [reply pkg/applicationmanager/controllers/applicationrestore.go:727]: Fixed

File: drivers/volume/csi/csi.go [hunk 3]
```diff
@@ -552,8 +552,8 @@ func (c *csi) uploadObject(
 		return err
 	}
 
-	if backupLocation.Location.EncryptionKey != "" {
```
Comment (adityadani) [inline drivers/volume/csi/csi.go:555]: Lets fail the uploadObject if `backupLocation.Location.EncryptionKey` is set. Return an error message
```
EncryptionKey is deprecated, use EncryptionKeyV2 instead.
```

File: drivers/volume/csi/csi.go [hunk 4]
```diff
@@ -1234,10 +1234,12 @@ func (c *csi) downloadObject(
 	if err != nil {
 		return nil, err
 	}
-	if restoreLocation.Location.EncryptionKey != "" {
```
Comment (adityadani) [inline drivers/volume/csi/csi.go:1237]: same for downloadObject

File: drivers/volume/csi/csi.go [hunk 5]
```diff
@@ -1234,10 +1234,12 @@ func (c *csi) downloadObject(
...
-		if data, err = crypto.Decrypt(data, restoreLocation.Location.EncryptionKey); err != nil {
-			return nil, err
+	if restoreLocation.Location.EncryptionV2Key != "" {
+		var decryptData []byte
+		if decryptData, err = crypto.Decrypt(data, restoreLocation.Location.EncryptionV2Key); err != nil {
+			return data, nil
```
Comment (adityadani) [inline drivers/volume/csi/csi.go:1240]: nit: add a log line indicating why decrypt failed and us returning the data as is

File: pkg/applicationmanager/controllers/applicationbackup.go [hunk 6]
```diff
@@ -935,8 +935,8 @@ func (a *ApplicationBackupController) uploadObject(
 		return err
 	}
 
-	if backupLocation.Location.EncryptionKey != "" {
-		if data, err = crypto.Encrypt(data, backupLocation.Location.EncryptionKey); err != nil {
+	if backupLocation.Location.EncryptionV2Key != "" {
```
Comment (adityadani) [inline pkg/applicationmanager/controllers/applicationbackup.go:938]: same here

File: pkg/applicationmanager/controllers/applicationrestore.go [hunk 7]
```diff
@@ -721,10 +721,13 @@ func (a *ApplicationRestoreController) downloadObject(
 	if err != nil {
 		return nil, err
 	}
-	if restoreLocation.Location.EncryptionKey != "" {
```
Comment (adityadani) [inline pkg/applicationmanager/controllers/applicationrestore.go:724]: same here: return an error

File: pkg/applicationmanager/controllers/backupsync.go [hunk 8]
```diff
@@ -101,10 +101,11 @@ func (b *BackupSyncController) syncBackupsFromLocation(location *storkv1.BackupL
 					log.BackupLocationLog(location).Errorf("Error syncing backup %v: %v", backupName, err)
 					continue
 				}
-				if location.Location.EncryptionKey != "" {
```
Comment (adityadani) [inline pkg/applicationmanager/controllers/backupsync.go:104]: here too

File: pkg/crypto/crypto.go [hunk 9]
```diff
@@ -29,6 +29,11 @@ func Decrypt(data []byte, passphrase string) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Sometime, if we pass uncrypted data, which does not have header,
```
Comment (adityadani) [inline pkg/crypto/crypto.go:32]:
```suggestion
	// Sometime, if we pass unencrypted data, which does not have header,
```

File: pkg/snapshotter/snapshotter_csi.go [hunk 10]
```diff
@@ -1032,8 +1032,8 @@ func (c *csiDriver) UploadSnapshotObjects(
 		return err
 	}
 
-	if backupLocation.Location.EncryptionKey != "" {
```
Comment (adityadani) [inline pkg/snapshotter/snapshotter_csi.go:1035]: return errors in this file too

**Review by adityadani — APPROVED** [review]

File: drivers/volume/csi/csi.go (hunk 3 above)
Comment (siva-portworx) [reply drivers/volume/csi/csi.go:555]: Fixed

File: drivers/volume/csi/csi.go (hunk 4 above)
Comment (siva-portworx) [reply drivers/volume/csi/csi.go:1237]: Fixed

File: drivers/volume/csi/csi.go (hunk 5 above)
Comment (siva-portworx) [reply drivers/volume/csi/csi.go:1240]: fixed

File: pkg/applicationmanager/controllers/applicationrestore.go (hunk 7 above)
Comment (siva-portworx) [reply pkg/applicationmanager/controllers/applicationrestore.go:724]: Fixed.

File: pkg/applicationmanager/controllers/backupsync.go (hunk 8 above)
Comment (siva-portworx) [reply pkg/applicationmanager/controllers/backupsync.go:104]: fixed

File: pkg/snapshotter/snapshotter_csi.go (hunk 10 above)
Comment (siva-portworx) [reply pkg/snapshotter/snapshotter_csi.go:1035]: fixed.

File: pkg/applicationmanager/controllers/applicationbackup.go (hunk 6 above)
Comment (siva-portworx) [reply pkg/applicationmanager/controllers/applicationbackup.go:938]: Fixed

**Pull request merged**